ros/src/mas_perception_libs/scene_detection_action.py

```diff
@@ -58,7 +60,142 @@ def is_object_on_plane(plane_msg, obj_box_msg, z_tolerance=0.05):
+
+    def _execute_cb(self, _):
+        # subscribe and wait for cloud message
+        print('Object detection action request received')
+        print('Waiting for cloud message...')
+        try:
+            cloud_msg = rospy.wait_for_message(self._cloud_topic, PointCloud2,
+                                               timeout=self._timeout_s)
+        except rospy.ROSInterruptException:
+            # shutdown event interrupts waiting for point cloud
+            raise
+        except rospy.ROSException:
+            rospy.logwarn('timeout waiting for cloud topic "{0}" after "{1}" seconds'
+                          .format(self._cloud_topic, self._timeout_s))
+            self._action_server.set_aborted()
+            return
+
+        rospy.loginfo('Cloud message received; detecting objects...')
+        img_msg = cloud_msg_to_image_msg(cloud_msg)
+        try:
+            bounding_boxes, classes, confidences = self._detector_handler.process_image_msg(img_msg)
+        except RuntimeError as e:
+            self._action_server.set_aborted(text=e.message)
+            return
+
+        rospy.loginfo('transforming cloud to frame: ' + self._target_frame)
+        try:
+            transformed_cloud_msg = transform_cloud_with_listener(cloud_msg, self._target_frame, self._tf_listener)
+        except RuntimeError as e:
+            self._action_server.set_aborted(text=e.message)
+            return
+
+        if self._filtered_cloud_pub.get_num_connections() > 0:
+            self._filtered_cloud_pub.publish(filtered_cloud)
+
+        rospy.loginfo('creating action result and setting success')
+        result = PlaneDetectionActionServer._get_action_result(transformed_cloud_msg, bounding_boxes,
+                                                               classes, confidences)
+        self._action_server.set_succeeded(result)
+
+    @staticmethod
+    def _get_action_result(cloud_msg, bounding_boxes, classes, confidences):
+        """
+        :type cloud_msg: PointCloud2
+        :type bounding_boxes: list
+        :type classes: list
+        :type confidences: list
+        :rtype: DetectObjectsResult
+        """
+        result = DetectObjectsResult()
+        for index, box in enumerate(bounding_boxes):
+            # TODO(minhnh): BoundingBox code returns bad positions and dimensions
+            # check if object in on plane
+            cropped_cloud = crop_organized_cloud_msg(cloud_msg, box)
+
+            obj_coords = crop_cloud_to_xyz(cropped_cloud, BoundingBox2D(box_geometry=(0, 0, cropped_cloud.width,
+                                                                                      cropped_cloud.height)))
+            obj_coords = np.reshape(obj_coords, (-1, 3))
+            min_coord = np.nanmin(obj_coords, axis=0)
+            max_coord = np.nanmax(obj_coords, axis=0)
+            mean_coord = np.nanmean(obj_coords, axis=0)
+
+            # fill object geometry info
```

Comment:
could move this to a different function to reduce repetitive code

Reply:
Added a function in `utils` for this.